[MPS][EZ] Fix cfloat->chalf conversion on MacOS13

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119670
* #119669
* __->__ #119681

By using `view_as_real` when type casting between two complex types

malfet committed:
[MPS][EZ] Fix cfloat->chalf conversion on MacOS13

By using `view_as_real` when type casting between two complex types

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/mps/operations/Copy.mm:
```diff
@@ -271,8 +272,12 @@ void copy_blit_mps(void* dst, const void* src, size_t size) {
-      at::real(dst_) = src;
-      at::imag(dst_) = at::zeros_like(src);
+      if (!src.is_complex()) {
+        at::real(dst_) = src;
```
This kind of assignment doesn't really do anything? do you mean `at::real(dst_).copy_(src)` ?

malfet replied:
Not sure I understand what do you mean. It copies (with cast) values of src into real part of `dst`, isn't it?

I.e. on MacOS13:
```
% python -c "import torch;print(torch.arange(10, device='mps').cfloat())"
tensor([0.+0.j, 1.+0.j, 2.+0.j, 3.+0.j, 4.+0.j, 5.+0.j, 6.+0.j, 7.+0.j, 8.+0.j, 9.+0.j],
       device='mps:0')
```
If it didn't work, I've assumed it would yield garbage, would't it?

ezyang replied:
we have a special operator= on tensor that makes this work

malfet replied:
Changed it to use `copy_` as it's more explicit this way

albanD commented on aten/src/ATen/native/mps/operations/Copy.mm:
```diff
@@ -271,8 +272,12 @@ void copy_blit_mps(void* dst, const void* src, size_t size) {
-      at::imag(dst_) = at::zeros_like(src);
+      if (!src.is_complex()) {
+        at::real(dst_) = src;
+        at::imag(dst_) = at::zeros_like(src);
```
```suggestion
        at::imag(dst_).fill_(0);
```
That will avoid creating a full size Tensor with zeros

malfet committed:
Update on "[MPS][EZ] Fix cfloat->chalf conversion on MacOS13"


By using `view_as_real` when type casting between two complex types

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

malfet commented:
@pytorchbot merge -f "MPS tests are green"
